### PR TITLE
C51-380: "Create Email" Button Functionality fix

### PR DESCRIPTION
**Pull request by deb1990**
## Overview
As part of this ticket the "Create Email" Button Functionality has been fixed.

## Before
before

## After
after

## Technical Details
1. The url parameters to open the Create Email popup was wrong. It should be same as the New "Email" Activity. Which has been corrected
`ng-href="{{ 'civicrm/activity/email/add' | civicaseCrmUrl:{ action: 'add', caseid: item.id, atype: '3', reset: 1, context: 'standalone' } }}"`

2. `Actions.js` has been renamed to `CaseActions.js`.

## Additional Fixes
Also it was found that the Create PDF Letter functionality was not working correctly before.

### Before
before

### After
after

### Technical Details
The functionality should be same as `Print/merge Document`. So new service `PrintMergeCaseAction ` has been created to share the code.

### CODE REVIEW

File: ang/test/civicase/Actions--print-merge.spec.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    describe('getPath()', function () {
+      var caseObj, returnValue;
+
+      beforeEach(function () {
+        caseObj = CasesMockData.get().values[0];
+
+        returnValue = PrintMergeCaseAction.getPath([caseObj]);
+      });
+
+      it('returns path to open popup for sending printing/merging', function () {
```
Comment (swastikpareek) [inline ang/test/civicase/Actions--print-merge.spec.js]: The statement `returns path to open popup for sending printing/merging` seems a bit confusing. Do you think we should make it `returns path to open popup for sending the activity for printing/merging` ? It's not a big deal and we can keep it as it too. Your call.

File: ang/civicase/Actions--print-merge.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    this.getPath = function (cases) {
+      var contactIds = [];
+      var caseIds = [];
+
+      _.each(cases, function (item) {
+        caseIds.push(item.id);
+        contactIds.push(item.client[0].contact_id);
+      });
+
+      var popupPath = {
```
Comment (swastikpareek) [inline ang/civicase/Actions--print-merge.js]: Just wondering. Can't we just return the object directly, without saving it in a var popupPath?
```js
return {
  path: 'civicrm/activity/pdf/add',
  query: {
    action: 'add',
    reset: 1,
    context: 'standalone',
    cid: contactIds.join(),
    caseid: caseIds.join()
  }
};
```
In this way we save an extra variable declaration and put the main focus on the logic and not extra piece of code.
And we can make up for readability of the function by adding extra line to function docs like this.
```js
    /** 
     * Get path object to print/merge the activity
     *
     * @param {Array} cases
     * @return {Object} popup path object
     */

File: ang/civicase/Actions.html [hunk 3]
```diff
@@ -0,0 +1,8 @@
+<li ng-class="{disabled: !isActionEnabled(action)}" ng-if="isActionAllowed(action)" ng-repeat="action in caseActions">
+  <a href ng-click="doAction(action)"><i class="fa {{action.icon}}"></i> {{ action.title }}</a>
```
Comment (swastikpareek) [inline ang/civicase/Actions.html]: Can we indent it properly?
```html
<a href ng-click="doAction(action)">
  <i class="fa {{action.icon}}"></i> {{ action.title }}
</a>

File: ang/civicase/Actions.html [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+  <ul ng-if="hasSubMenu(action)" class="dropdown-menu sub-menu">
+    <li ng-class="{disabled: !isActionEnabled(subMenu)}" ng-if="isActionAllowed(subMenu)" ng-repeat="subMenu in action.items">
+      <a href ng-click="doAction(subMenu)"><i class="fa {{subMenu.icon}}"></i> {{ subMenu.title }}</a>
```
Comment (swastikpareek) [inline ang/civicase/Actions.html]: Same as above.

File: ang/civicase/Actions.html [hunk 5]
```diff
@@ -0,0 +1,8 @@
+<li ng-class="{disabled: !isActionEnabled(action)}" ng-if="isActionAllowed(action)" ng-repeat="action in caseActions">
+  <a href ng-click="doAction(action)"><i class="fa {{action.icon}}"></i> {{ action.title }}</a>
+  <ul ng-if="hasSubMenu(action)" class="dropdown-menu sub-menu">
+    <li ng-class="{disabled: !isActionEnabled(subMenu)}" ng-if="isActionAllowed(subMenu)" ng-repeat="subMenu in action.items">
```
Comment (swastikpareek) [inline ang/civicase/Actions.html]: Can we can bring the attributes in new line for readability?
```html
<li ng-class="{disabled: !isActionEnabled(subMenu)}"
     ng-if="isActionAllowed(subMenu)" 
     ng-repeat="subMenu in action.items">
....
....
</li>
```

File: ang/civicase/Actions.js [hunk 6]
```diff
@@ -1,386 +1,364 @@
...
-              };
+      link: civicaseActionsLink
+    };
 
-              _.each(CRM.civicase.tagsets, function (tagset) {
-                model[tagset.id] = [];
-                keys.push(tagset.id);
-              });
+    /**
+     * Angular JS's link function for civicaseActions Directive
```
Comment (swastikpareek) [inline ang/civicase/Actions.js]: Please add an empty line after function description and before params
```js
 /**
  * Angular JS's link function for civicaseActions Directive
  *
  * @param {Object} $scope
  * @param {Object} element
  * @param {Object} attributes
  */
```

File: ang/civicase/Actions.js [hunk 7]
```diff
@@ -1,386 +1,364 @@
...
+     * @param {Object} $scope
+     * @param {Object} element
+     * @param {Object} attributes
+     */
+    function civicaseActionsLink ($scope, element, attributes) {
+      var ts = CRM.ts('civicase');
+      var multi = $scope.multi = attributes.multiple;
```
Comment (swastikpareek) [inline ang/civicase/Actions.js]: Can't we just save `attribute.multiple` in `multi` var and save `$scope.multi`, as we are not using this variable in the template ( I guess)
```
var multi = attributes.multiple;
```

Comment (swastikpareek) [reply ang/civicase/Actions.js]: Also, if we can have better naming for this var for improved readability?
`supportsMultipleValues` maybe ?

File: ang/civicase/Actions.js [hunk 8]
```diff
@@ -1,386 +1,364 @@
...
+
+            dialogService.open('EditTags', '~/civicase/EditTags.html', model, {
+              autoOpen: false,
+              height: 'auto',
+              width: '40%',
+              title: action.title,
+              buttons: [{
+                text: ts('Save'),
+                icons: {primary: 'fa-check'},
+                click: function () {
```
Comment (swastikpareek) [inline ang/civicase/Actions.js]: Can we encapsulate click function code in some other private function
```
function editTagModalClickEvent () {
.....
....
}
```
and 
```
buttons: [{
  text: ts('Save'),
  icons: {primary: 'fa-check'},
  click: editTagModalClickEvent
}]
```

File: ang/civicase/Actions.js [hunk 9]
```diff
@@ -1,386 +1,364 @@
...
-                      return params;
-                    }
-
-                    _.each(keys, function (key) {
-                      _.each(model[key], function (id) {
-                        values.push(id);
-                      });
+                  function tagParams (tagIds) {
+                    var params = {entity_id: item.id, entity_table: 'civicrm_case'};
+                    _.each(tagIds, function (id, i) {
```
Comment (swastikpareek) [inline ang/civicase/Actions.js]: Add a new line after var declarations.

File: ang/civicase/Actions.js [hunk 10]
```diff
@@ -1,386 +1,364 @@
...
+                      params['tag_id_' + i] = id;
                     });
+                    return params;
```
Comment (swastikpareek) [inline ang/civicase/Actions.js]: Add new line before `return` statement

File: ang/civicase/Actions.js [hunk 11]
```diff
@@ -1,386 +1,364 @@
...
+                  if (toAdd.length) {
+                    calls.push(['EntityTag', 'create', tagParams(toAdd)]);
+                  }
 
-                default:
-                  msg = cases.length === 1 ? ts('This case and all associated activities will be moved to the trash.') : ts('%1 cases and all associated activities will be moved to the trash.', {'1': cases.length});
-                  mode = 'delete';
-              }
+                  if (calls.length) {
+                    calls.push(['Activity', 'create', {case_id: item.id, status_id: 'Completed', activity_type_id: 'Change Case Tags'}]);
```
Comment (swastikpareek) [inline ang/civicase/Actions.js]: Could we change this to better readable form 
```js
calls.push(['Activity', 'create', {
  case_id: item.id, 
  status_id: 'Completed', 
  activity_type_id: 'Change Case Tags'
}]);
```

File: ang/civicase/Actions.js [hunk 12]
```diff
@@ -1,386 +1,364 @@
...
+
+              default:
+                msg = cases.length === 1 ? ts('This case and all associated activities will be moved to the trash.') : ts('%1 cases and all associated activities will be moved to the trash.', {'1': cases.length});
+                mode = 'delete';
+            }
 
-                  $scope.refresh(calls);
+            CRM.confirm({title: action.title, message: msg})
+              .on('crmConfirm:yes', function () {
+                var calls = [];
```
Comment (swastikpareek) [inline ang/civicase/Actions.js]: Add a new line after `var` declarations

File: ang/civicase/Actions.js [hunk 13]
```diff
@@ -1,386 +1,364 @@
...
-                    });
-                    $scope.refresh(calls);
-                  }
+          },
+
+          changeStatus: function (cases) {
+            var types = _.uniq(_.map(cases, 'case_type_id'));
+            var currentStatuses = _.uniq(_.collect(cases, 'status_id'));
+            var currentStatus = currentStatuses.length === 1 ? currentStatuses[0] : null;
+            var msg = '<form>' +
```
Comment (swastikpareek) [inline ang/civicase/Actions.js]: Just wondering. can we use template string literals here
```js
 var msg = `<form>
              <div>
                <input name="change_case_status" placeholder="${ts('Select New Status')}" />
              </div>
              <label for="change_case_status_details">${ts('Notes')} </label>
              <textarea id="change_case_status_details"></textarea>
            </form>`;
```

File: ang/civicase/Actions.js [hunk 14]
```diff
@@ -1,386 +1,364 @@
...
+                '<div><input name="change_case_status" placeholder="' + ts('Select New Status') + '" /></div>' +
+                '<label for="change_case_status_details">' + ts('Notes') + '</label>' +
+                '<textarea id="change_case_status_details"></textarea>' +
+                '</form>';
+            var statuses = _.map(CRM.civicase.caseStatuses, function (item, statusId) {
+              return { id: item.name, text: item.label, disabled: statusId === currentStatus };
```
Comment (swastikpareek) [inline ang/civicase/Actions.js]: Indent it for more readability
```js
return { 
  id: item.name, 
  text: item.label, 
  disabled: statusId === currentStatus 
};
```

File: ang/civicase/Actions.js [hunk 15]
```diff
@@ -1,386 +1,364 @@
...
+            });
+
+            _.each(types, function (caseTypeId) {
+              var allowedStatuses = CRM.civicase.caseTypes[caseTypeId].definition.statuses || [];
+              if (allowedStatuses.length) {
```
Comment (swastikpareek) [inline ang/civicase/Actions.js]: Add a new line after `var`.

File: ang/civicase/Actions.js [hunk 16]
```diff
@@ -1,386 +1,364 @@
...
+              title: action.title,
+              message: msg,
+              open: function () {
+                $('input[name=change_case_status]', this).crmSelect2({data: statuses});
+                CRM.wysiwyg.create('#change_case_status_details').then(function () {
+                  alignDialogBoxCenter(dialog);
+                });
+              }
+            })
+              .on('crmConfirm:yes', function () {
```
Comment (swastikpareek) [inline ang/civicase/Actions.js]: Check the indentations. Right now it's 7 . Should be 6( In indent with `dialog` var. Please let me know if there is any confusion on this.

File: ang/civicase/Actions.js [hunk 17]
```diff
@@ -1,386 +1,364 @@
...
-              _.each(cases, function (item) {
-                if (item.manager) {
-                  managers.push(item.manager.contact_id);
+                if (status) {
+                  _.each(cases, function (item) {
+                    var subject = ts('Case status changed from %1 to %2', {
+                      1: item.status,
+                      2: _.result(_.find(statuses, {id: status}), 'text')
+                    });
+                    calls.push(['Case', 'create', {id: item.id, status_id: status}]);
```
Comment (swastikpareek) [inline ang/civicase/Actions.js]: Add a new line after `var` declarations

File: ang/civicase/Actions.js [hunk 18]
```diff
@@ -1,386 +1,364 @@
...
+                    calls.push(['Activity', 'create', {case_id: item.id, status_id: 'Completed', activity_type_id: 'Change Case Status', subject: subject, details: details}]);
+                  });
```
Comment (swastikpareek) [inline ang/civicase/Actions.js]: Add a new line after `_.each`.

File: ang/civicase/Actions.js [hunk 19]
```diff
@@ -1,386 +1,364 @@
...
+          exportCases: function (cases) {
+            var caseIds = _.collect(cases, 'id');
+            var popupPath = {
+              path: 'civicrm/export/standalone',
+              query: {
+                reset: 1,
+                entity: 'Case',
+                id: caseIds.join()
+              }
+            };
```
Comment (swastikpareek) [inline ang/civicase/Actions.js]: Add a new line after `var` declarations

File: ang/civicase/Actions.js [hunk 20]
```diff
@@ -1,386 +1,364 @@
...
+
+          print: function (selectedCase) {
+            var url = CRM.url('civicrm/case/report/print', {
+              all: 1,
+              redact: 0,
+              cid: selectedCase.client[0].contact_id,
+              asn: 'standard_timeline',
+              caseID: selectedCase.id
+            });
+            var win = window.open(url, '_blank');
```
Comment (swastikpareek) [inline ang/civicase/Actions.js]: add a new line after `var`

File: ang/civicase/Actions.js [hunk 21]
```diff
@@ -1,386 +1,364 @@
...
+            win.focus();
+          },
+
+          gotoWebform: function (selectedCase, path, clientId) {
+            clientId = 'cid' + clientId;
+            var url = CRM.url(path, {
+              case1: selectedCase.id,
+              [clientId]: selectedCase.client[0].contact_id
+            });
+            var win = window.open(url, '_blank');
```
Comment (swastikpareek) [inline ang/civicase/Actions.js]: Same as above

File: ang/civicase/Actions.js [hunk 22]
```diff
@@ -1,386 +1,364 @@
...
+            win.focus();
+          },
+
+          lockCases: function (currentCase) {
```
Comment (swastikpareek) [inline ang/civicase/Actions.js]: Can't we directly return the array and add the info in the function docs.
All the functions lacks docs. Please add docs to each function

File: ang/civicase/Actions.js [hunk 23]
```diff
@@ -1,386 +1,364 @@
...
+      $scope.$watchCollection('cases', function (cases) {
+        // Special actions when viewing deleted cases
+        if (cases.length && cases[0].is_deleted) {
+          $scope.caseActions = [
+            {action: 'deleteCases(cases, "delete")', title: ts('Delete Permanently')},
+            {action: 'deleteCases(cases, "restore")', title: ts('Restore from Trash')}
+          ];
+        } else {
+          $scope.caseActions = _.cloneDeep(CRM.civicase.caseActions);
+          if (!$scope.multi) {
```
Comment (swastikpareek) [inline ang/civicase/Actions.js]: Please add a new line before `if` statement

File: ang/test/civicase/Actions--print-merge.spec.js (hunk 1 above)
Comment (deb1990) [reply ang/test/civicase/Actions--print-merge.spec.js]: I think `returns path for opening popup to print/merge document` is more appropriate. what do you think? thanks for the pointing out the issue :)

File: ang/civicase/Actions.js (hunk 7 above)
Comment (deb1990) [reply ang/civicase/Actions.js]: Renamed to `isBulkMode`. Also removed the `$scope` variable as it was not being used.

File: ang/civicase/Actions.js (hunk 13 above)
Comment (deb1990) [reply ang/civicase/Actions.js]: That is ES6 which we are not using at the moment.

File: ang/civicase/Actions.js (hunk 22 above)
Comment (deb1990) [reply ang/civicase/Actions.js]: This whole file was refactored to fix the linting issues, and also to apply our styleguide. But as I have not worked specifically on this file/functionality, I am not aware what each function exactly does, so I am skipping any kind of documentation/major refactoring.

File: ang/test/civicase/Actions--print-merge.spec.js (hunk 1 above)
Comment (swastikpareek) [reply ang/test/civicase/Actions--print-merge.spec.js]: Looks good to me too

**Pull request merged**